File: tools/tri_language_state_runner.ml
(* tools/tri_language_state_runner.ml
   Cross-language state runner for OCaml (Hermes Engine) communicating with
   Gleam (BEAM ETS) and Mojo (MAX SIMD) via Zenoh (port 8080/7447) and Wisp (port 4100).
   STAMP: SC-GLM-UI-001, SC-ZMOF-001, SC-CHECKLIST-001
*)

let zenoh_base = "http://127.0.0.1:8080"
let gleam_wisp_base = "http://127.0.0.1:4100"

let exec_cmd cmd =
  let ic = Unix.open_process_in cmd in
  let buf = Buffer.create 1024 in
  (try
     while true do
       let line = input_line ic in
       Buffer.add_string buf line;
       Buffer.add_char buf '\n'
     done
   with End_of_file -> ());
  let status = Unix.close_process_in ic in
  (status, Buffer.contents buf)

let http_get url =
  let cmd = Printf.sprintf "curl -s -m 5 '%s'" url in
  let (status, output) = exec_cmd cmd in
  match status with
  | Unix.WEXITED 0 -> Some output
  | _ -> None

let http_put url body =
  let cmd = Printf.sprintf "curl -s -m 5 -X PUT -d '%s' '%s'" body url in
  let (status, output) = exec_cmd cmd in
  match status with
  | Unix.WEXITED 0 -> Some output
  | _ -> None

let string_contains haystack needle =
  let h_len = String.length haystack in
  let n_len = String.length needle in
  if n_len = 0 then true
  else if h_len < n_len then false
  else
    let found = ref false in
    for i = 0 to h_len - n_len do
      if not !found && String.sub haystack i n_len = needle then
        found := true
    done;
    !found

let publish_zenoh_state key value =
  let url = Printf.sprintf "%s/c3i/a2a/ets/%s" zenoh_base key in
  Printf.printf "[OCAML-HERMES] Publishing to Zenoh: %s -> %s\n%!" url value;
  match http_put url value with
  | Some _ ->
      Printf.printf "[OCAML-HERMES] Successfully published '%s' to Zenoh\n%!" key;
      true
  | None ->
      Printf.eprintf "[OCAML-HERMES] Failed to publish '%s' to Zenoh\n%!" key;
      false

let get_zenoh_state key =
  let url = Printf.sprintf "%s/c3i/a2a/ets/%s" zenoh_base key in
  http_get url

let put_ets_state key value =
  let url = Printf.sprintf "%s/api/v1/ets/put?key=%s&val=%s" gleam_wisp_base key value in
  Printf.printf "[OCAML-HERMES] Putting to BEAM ETS via Wisp: %s\n%!" url;
  match http_get url with
  | Some out ->
      Printf.printf "[OCAML-HERMES] ETS put response: %s\n%!" (String.trim out);
      true
  | None ->
      Printf.eprintf "[OCAML-HERMES] ETS put failed for %s\n%!" key;
      false

let get_ets_state key =
  let url = Printf.sprintf "%s/api/v1/ets/%s" gleam_wisp_base key in
  http_get url

let get_tri_language_state () =
  let url = Printf.sprintf "%s/api/v1/state/tri_language" gleam_wisp_base in
  http_get url

let run_test () =
  Printf.printf "========================================================\n%!";
  Printf.printf "[OCAML-HERMES] Starting Tri-Language Zenoh & ETS Test...\n%!";
  Printf.printf "========================================================\n%!";

  (* 1. Publish OCaml state to Zenoh *)
  let ocaml_payload = "OCAML_HERMES_ORACLE_ACTIVE" in
  let pub_ok = publish_zenoh_state "ocaml_state" ocaml_payload in
  assert pub_ok;

  (* 2. Put OCaml state directly to ETS as well for bidirectional parity *)
  let ets_ok = put_ets_state "ocaml_state" ocaml_payload in
  assert ets_ok;

  (* 3. Read back from Zenoh *)
  (match get_zenoh_state "ocaml_state" with
   | Some content ->
       Printf.printf "[OCAML-HERMES] Readback from Zenoh: %s\n%!" (String.trim content);
       assert (String.length content > 0)
   | None ->
       failwith "Failed to read ocaml_state from Zenoh");

  (* 4. Read back from ETS *)
  (match get_ets_state "ocaml_state" with
   | Some content ->
       Printf.printf "[OCAML-HERMES] Readback from ETS: %s\n%!" (String.trim content);
       assert (string_contains content "OCAML_HERMES_ORACLE_ACTIVE")
   | None ->
       failwith "Failed to read ocaml_state from ETS");

  (* 5. Verify Gleam state in ETS / Zenoh *)
  (match get_ets_state "gleam_state" with
   | Some content ->
       Printf.printf "[OCAML-HERMES] Observed Gleam state in ETS: %s\n%!" (String.trim content);
       assert (string_contains content "GLEAM_OTP29_SUPERVISOR_ACTIVE")
   | None ->
       Printf.printf "[OCAML-HERMES] Warning: gleam_state not yet in ETS\n%!");

  (* 6. Read tri-language state summary *)
  (match get_tri_language_state () with
   | Some content ->
       Printf.printf "[OCAML-HERMES] Tri-Language State Summary: %s\n%!" (String.trim content);
       assert (string_contains content "GLEAM_OTP29_SUPERVISOR_ACTIVE");
       assert (string_contains content "OCAML_HERMES_ORACLE_ACTIVE")
   | None ->
       failwith "Failed to fetch /api/v1/state/tri_language");

  (* 7. Publish fractal telemetry hooks to Zenoh & ETS *)
  let now = Unix.gettimeofday () in
  let telem_payload = Printf.sprintf "{\"subsystem\":\"Hermes_Engine\",\"language\":\"OCaml\",\"status\":\"PASSED\",\"fractal_layer\":\"L3_TRANSACTION\",\"timestamp\":%.3f}" now in
  let _ = http_put (Printf.sprintf "%s/c3i/testing/events/ocaml" zenoh_base) telem_payload in
  let _ = http_put (Printf.sprintf "%s/indrajaal/otel/ops/testing/ocaml" zenoh_base) telem_payload in
  let _ = put_ets_state "test:ocaml:status" "PASSED" in
  let _ = put_ets_state "test:ocaml:timestamp" (Printf.sprintf "%.3f" now) in
  let _ = put_ets_state "test:ocaml:runner" "tools/tri_language_state_runner.ml" in
  Printf.printf "[OCAML-HERMES] Telemetry hooks published to Zenoh (c3i/testing/events/ocaml) and ETS (test:ocaml:status)\n%!";

  Printf.printf "========================================================\n%!";
  Printf.printf "[OCAML-HERMES] ALL OCAML ZENOH/ETS PARITY CHECKS PASSED!\n%!";
  Printf.printf "========================================================\n%!"

let () =
  try
    run_test ();
    exit 0
  with exn ->
    Printf.eprintf "[OCAML-HERMES] Test encountered exception: %s\n%!" (Printexc.to_string exn);
    exit 1

File: services/inference/max/tri_language_state_runner.py
#!/usr/bin/env python3
"""
[C3I-SIL6-MSTS] UOS TRI-LANGUAGE ZENOH & ETS STATE RUNNER (MODULAR MAX / MOJO TIER)
STAMP: SC-GLM-UI-001, SC-ZMOF-001, SC-COG-001, SC-COG-MAX-001, CHK-07-DRIVE

Enforces cross-language state sharing and consensus between Gleam (BEAM),
OCaml (Hermes), and Mojo/Python (MAX SIMD) via Zenoh pub/sub mesh (8080/7447)
and BEAM ETS (port 4100).
"""

import sys
import json
import urllib.request
import urllib.parse

HARD_DENIED_SYSTEM_OS_SERIAL = "25503L801736"
ZENOH_BASE = "http://127.0.0.1:8080"
GLEAM_WISP_BASE = "http://127.0.0.1:4100"

def assert_storage_safety(payload_str: str):
    """Safety interlock check (CHK-07-DRIVE): Ensure root OS NVMe drive serial is protected."""
    if HARD_DENIED_SYSTEM_OS_SERIAL in payload_str:
        raise PermissionError(f"HARD_DENIED: root OS NVMe drive {HARD_DENIED_SYSTEM_OS_SERIAL} locked against mutation")

def http_get(url: str) -> str:
    assert_storage_safety(url)
    req = urllib.request.Request(url, headers={"User-Agent": "UOS-MAX-Mojo-Runner/1.0"})
    with urllib.request.urlopen(req, timeout=5) as resp:
        return resp.read().decode("utf-8")

def http_put(url: str, body: str) -> str:
    assert_storage_safety(url)
    assert_storage_safety(body)
    data = body.encode("utf-8")
    req = urllib.request.Request(url, data=data, method="PUT", headers={
        "Content-Type": "text/plain",
        "User-Agent": "UOS-MAX-Mojo-Runner/1.0"
    })
    with urllib.request.urlopen(req, timeout=5) as resp:
        return resp.read().decode("utf-8")

def publish_zenoh_state(key: str, value: str):
    url = f"{ZENOH_BASE}/c3i/a2a/ets/{key}"
    print(f"[MOJO-MAX] Publishing to Zenoh: {url} -> {value}")
    resp = http_put(url, value)
    print(f"[MOJO-MAX] Zenoh publish response: {resp}")
    return True

def get_zenoh_state(key: str):
    url = f"{ZENOH_BASE}/c3i/a2a/ets/{key}"
    return http_get(url)

def put_ets_state(key: str, value: str):
    encoded_key = urllib.parse.quote(key)
    encoded_val = urllib.parse.quote(value)
    url = f"{GLEAM_WISP_BASE}/api/v1/ets/put?key={encoded_key}&val={encoded_val}"
    print(f"[MOJO-MAX] Putting to BEAM ETS via Wisp: {url}")
    resp = http_get(url)
    print(f"[MOJO-MAX] Wisp ETS put response: {resp.strip()}")
    return True

def get_ets_state(key: str):
    url = f"{GLEAM_WISP_BASE}/api/v1/ets/{urllib.parse.quote(key)}"
    return http_get(url)

def get_tri_language_state():
    url = f"{GLEAM_WISP_BASE}/api/v1/state/tri_language"
    return http_get(url)

def run_test():
    print("=" * 60)
    print("[MOJO-MAX] Starting Tri-Language Zenoh & ETS Test...")
    print("=" * 60)

    mojo_payload = "MOJO_MAX_SIMD_RANKER_ACTIVE"

    # 1. Publish Mojo state to Zenoh mesh
    publish_zenoh_state("mojo_state", mojo_payload)

    # 2. Put Mojo state into BEAM ETS
    put_ets_state("mojo_state", mojo_payload)

    # 3. Read back from Zenoh
    zenoh_val = get_zenoh_state("mojo_state")
    print(f"[MOJO-MAX] Readback from Zenoh: {zenoh_val.strip()}")
    assert mojo_payload in zenoh_val, f"Expected {mojo_payload} in Zenoh response"

    # 4. Read back from ETS
    ets_val = get_ets_state("mojo_state")
    print(f"[MOJO-MAX] Readback from ETS: {ets_val.strip()}")
    ets_json = json.loads(ets_val)
    assert ets_json.get("value") == mojo_payload, f"Expected {mojo_payload} in ETS response"

    # 5. Read Gleam state from ETS
    gleam_val = get_ets_state("gleam_state")
    print(f"[MOJO-MAX] Observed Gleam state in ETS: {gleam_val.strip()}")
    gleam_json = json.loads(gleam_val)
    assert gleam_json.get("value") == "GLEAM_OTP29_SUPERVISOR_ACTIVE"

    # 6. Read OCaml state from ETS
    ocaml_val = get_ets_state("ocaml_state")
    print(f"[MOJO-MAX] Observed OCaml state in ETS: {ocaml_val.strip()}")
    ocaml_json = json.loads(ocaml_val)
    assert ocaml_json.get("value") == "OCAML_HERMES_ORACLE_ACTIVE"

    # 7. Check full tri-language convergence
    tri_summary_str = get_tri_language_state()
    print(f"[MOJO-MAX] Tri-Language State Summary: {tri_summary_str.strip()}")
    tri_json = json.loads(tri_summary_str)

    assert tri_json.get("gleam_state") == "GLEAM_OTP29_SUPERVISOR_ACTIVE", "Gleam state mismatch"
    assert tri_json.get("ocaml_state") == "OCAML_HERMES_ORACLE_ACTIVE", "OCaml state mismatch"
    assert tri_json.get("mojo_state") == "MOJO_MAX_SIMD_RANKER_ACTIVE", "Mojo state mismatch"
    assert tri_json.get("is_converged") is True, f"Tri-language mesh expected is_converged=True, got {tri_json.get('is_converged')}"

    print("=" * 60)
    print("[MOJO-MAX] FULL TRI-LANGUAGE CONVERGENCE VERIFIED!")
    print(f"[MOJO-MAX] Gleam (BEAM): {tri_json.get('gleam_state')}")
    print(f"[MOJO-MAX] OCaml (Hermes): {tri_json.get('ocaml_state')}")
    print(f"[MOJO-MAX] Mojo/Python (MAX): {tri_json.get('mojo_state')}")
    print(f"[MOJO-MAX] Convergence: {tri_json.get('is_converged')}")
    print(f"[MOJO-MAX] Total ETS Entries: {tri_json.get('ets_entry_count')}")
    print("=" * 60)

if __name__ == "__main__":
    try:
        run_test()
        sys.exit(0)
    except Exception as e:
        print(f"[MOJO-MAX] Test failed with exception: {e}", file=sys.stderr)
        sys.exit(1)
